Add new metric APIs (#25312)
dotnet/docs: docs/core/diagnostics/metrics-instrumentation.md

---
title: Creating Metrics
description: How to add new metrics to a .NET library or application
ms.topic: tutorial
ms.date: 11/04/2021
---

# Creating Metrics

> **This article applies to: ✔️** .NET Core 3.1 and later versions **✔️** .NET Framework 4.5 and later versions

Reviewer: ```suggestion
**This article applies to: ✔️** .NET Core 3.1 and later versions **✔️** .NET Framework 4.6.1 and later versions
```

the [metrics collection tutorial](metrics-collection.md) for other alternatives. If the [dotnet-counters](./dotnet-counters.md) tool is not already installed use the SDK
to install it:

```dotnetcli
> dotnet tool update -g dotnet-counters
You can invoke the tool using the following command: dotnet-counters
Tool 'dotnet-counters' (version '5.0.251802') was successfully installed.
```

While the sample app above is still running, list the running processes in a 2nd shell to determine the process ID:

```dotnetcli
> dotnet-counters ps
     10180 dotnet     C:\Program Files\dotnet\dotnet.exe
     19964 metric-instr E:\temp\metric-instr\bin\Debug\netcoreapp3.1\metric-instr.exe
```

Find the ID for the process name that matches the sample app and have dotnet-counters monitor the new counter:

```dotnetcli
> dotnet-counters monitor -p 19964 HatCo.HatStore
Press p to pause, r to resume, q to quit.
    Status: Running

[HatCo.HatStore]
    hats-sold (Count / 1 sec)                          4
```

As expected we can see that HatCo store is steadily selling 4 hats each second. 

### Types of Instruments

> Above the example created a <xref:System.Diagnostics.Metrics.Counter> instrument but there are several different types available. Instruments differ in two ways:

Reviewer 2: ???

```suggestion
The above example created a <xref:System.Diagnostics.Metrics.Counter> instrument but there are several different types available. Instruments differ in two ways:
```

Author: Your suggestion did capture my intended meaning. Not sure if your "???" meant there was something else you found confusing as well?

Reviewer 2: I found it hard to understand (perhaps due to my bad English). I interpreted it as "In the example we've only demonstrated the <xref:System.Diagnostics.Metrics.Counter> instrument, but there are more instrument types available."

Author: Your English is pretty sharp if you ask me ; ) Your phrasing was clearer so I adopted it.

- **Default metric computations** - Tools that collect and analyze the instrument measurements will compute different default metrics depending on the instrument.
- **Storage of aggregated data** - Most useful metrics need data to be aggregated from many measurements. One option is the caller provides individual measurements
  at arbitrary times and the collection tool will manage the aggregation. Alternately the caller can manage the aggregate measurements and provide them on-demand in a callback.

> Types of instruments currently available:

Reviewer 3: would it be good to give some examples when each instrument can be used. For example, Counter can be used for counting number of requests. ObservableGauge can be used to report room temperature,...etc.

Author: There is a section lower in the text that is 'Best practices selecting an instrument type' which does have this info. Do you think I should restructure to bring the examples closer, repeat the examples, or something else?

Author: @tarekgh - making sure you see the question here for you : )

Reviewer 3: I suggest bringing the section `Best practices selecting an instrument type` closer to this section. As the article reader, will give immediate idea why we have such instruments and when to use each one. I am ok if you feel otherwise too. this is just a suggestion :-)

Reviewer 2: FYI - we have some guidance regarding which instrument to be used in the OpenTelemetry spec repo https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/supplementary-guidelines.md#instrument-selection.

        Percentile=99                                146
    orders-pending                                     3
```

This example uses some randomly generated numbers so your values will vary a bit. We can see that hats-sold (the Counter) and
coats-sold (the ObservableCounter) both show up as a rate. The ObservableGauge, orders-pending, appears
as an absolute value. Dotnet-counters renders Histogram instruments as three percentile statistics (50th, 95th, and 99th) but other tools may
summarize the distribution differently or offer more configuration options.

#### Best practices selecting an instrument type

- For counting things, or any other value that solely increases over time, use Counter or ObservableCounter. Choose between Counter and ObservableCounter depending on which
  is easier to add to the existing code: either an API call for each increment operation or a callback that will read the current total from a variable the code maintains. In
  extremely hot code paths where performance is important and using <xref:System.Diagnostics.Metrics.Counter.Add%2A> would create >1 million calls/sec/thread, using
  ObservableCounter may offer more opportunity for optimization.

> - For timing things Histogram is usually prefered. Often it is useful to understand the tail of these distributions (90th, 95th, 99th percentile) rather than averages or

Reviewer 2: ```suggestion
- For timing things Histogram is usually preferred. Often it is useful to understand the tail of these distributions (90th, 95th, 99th percentile) rather than averages or
```

  totals.

- Other common cases such as business metrics, physical sensors, cache hit rates, or sizes of caches, queues, files, etc are usually well suited for ObservableGauge.

- Be aware that Histograms tend to store a lot more data in memory than other metric types however the exact memory usage is determined by the collection tool being used.

>   If you are defining a large number of Histogram metrics (>100) you may need to give users guidance not to enable them all at the same time or configure their tools to save

Reviewer 2: ```suggestion
  If you are defining a large number (>100) of Histogram metrics you may need to give users guidance not to enable them all at the same time or configure their tools to save
```

  instrument object, but in most cases you don't need to save it in a variable because no further interaction with the object is needed. Assigning it to a static variable
  as we did for the other instruments is legal but error prone because C# static initialization is lazy and the variable is usually never referenced. Here is an example
  of the problem:

```C#
using System;
using System.Diagnostics.Metrics;

class Program
{
    // BEWARE! Static initializers only run when code in a running method refers to a static variable. 
    // These statics will never be initialized because none of them were referenced in Main().
    //
    static Meter s_meter = new Meter("HatCo.HatStore", "1.0.0");
    static ObservableCounter<int> s_coatsSold = s_meter.CreateObservableCounter<int>("coats-sold", () => s_rand.Next(1,10));
    static Random s_rand = new Random();

    static void Main(string[] args)
    {
        Console.ReadLine();
    }
}
```

### Descriptions and Units

Instruments can specify optional descriptions and units. These values are opaque to all metric calculations but can be shown in collection tool UI

> to help engineers understand how to interpret the data. Stop the example process from above and replace the example code in Program.cs with:

Reviewer 2: ```suggestion
to help engineers understand how to interpret the data. Stop the example process from above and replace the example code in `Program.cs` with:
```

  then logs, transactional databases or big data processing systems may be more appropriate solutions to operate at the needed scale.

- For instruments that will have very large numbers of tag combinations, prefer using a smaller storage type to help reduce memory overhead. For example storing the `short` for
  a `Counter<short>` only occupies 2 bytes per tag combination whereas a `double` for `Counter<double>` occupies 8 bytes per tag combination.

- Collection tools are encouraged to optimize for code that specifies the same set of tag names in the same order for each call to record measurements on the
  same instrument. For high performance code that needs to call <xref:System.Diagnostics.Metrics.Counter`1.Add%2A> and <xref:System.Diagnostics.Metrics.Histogram`1.Record%2A>
  frequently prefer using the same sequence of tag names for each call.

- The .NET API is optimized to be allocation-free for <xref:System.Diagnostics.Metrics.Counter`1.Add%2A> and <xref:System.Diagnostics.Metrics.Histogram`1.Record%2A> calls

>   with three or fewer tags. In general the performance overhead of these calls increases as more tags are used.

Reviewer: mention `TagList` for alloc free above 3 as well.

Reviewer 3: TagList can help in avoid the allocation if having up to 8 tags :-) will have small cpu overhead for copying the tags.

            Thread.Sleep(100);
            s_hatsSold.Add(4);
        }
    }
}
```

Run the new process and use dotnet-counters as before in a 2nd shell to view the metrics:

```dotnetcli
Press p to pause, r to resume, q to quit.
    Status: Running

[HatCo.HatStore]
    hats-sold (Hats / 1 sec)                           40
```

dotnet-counters doesn't currently use the description text in the UI, but it does show the unit when it is provided. In this case we see "Hats"
has replaced the generic term "Count" that is visible in previous descriptions.

#### Best Practices

- The unit specified in the constructor should describe the units appropriate for an individual measurement. This will sometimes differ from the units on the final metric.
  In this example each measurement is a number of hats so "Hats" is appropriate unit to pass in the constructor. The collection tool calculated a rate and derived on it's own
  that the appropriate unit for the calculated metric is Hats/sec.

### Multi-dimensional metrics

Measurements can also be associated with key-value pairs called tags that allow data to be categorized for analysis. For example HatCo might want to record not

> only the number of hats that were sold, but also which size and color they were. Later when analyzing the data HatCo engineers can break out the totals by

Reviewer 4: ```suggestion
only the number of hats that were sold, but also which size and color they were. When analyzing the data later, HatCo engineers can break out the totals by
```

size, color, or any combination of both.

> Counter and Histogram tags can be specified in overloads of the <xref:System.Diagnostics.Metrics.Counter`1.Add%2A> and

Reviewer 4: ```suggestion
Counter and Histogram tags can be specified in overloads of the <xref:System.Diagnostics.Metrics.Counter%601.Add%2A> and
```

> <xref:System.Diagnostics.Metrics.Histogram`1.Record%2A> respectively that take one or more KeyValuePairs. For example:

Reviewer 4: ```suggestion
<xref:System.Diagnostics.Metrics.Histogram%601.Record%2A> that take one or more `KeyValuePair` arguments. For example:
```

```C#
s_hatsSold.Add(2,
               new KeyValuePair<string, object>("Color", "Red"),
               new KeyValuePair<string, object>("Size", 12));

```

> Replace the code of `Program.cs` and re-run the app and dotnet-counters as before:

Reviewer 4: ```suggestion
Replace the code of `Program.cs` and rerun the app and dotnet-counters as before:
```

            new Measurement<int>(1, new KeyValuePair<string,object>("Country", "Mexico")),
        };
    }
}
```

When run with dotnet-counters as before the result is:

```dotnetcli
Press p to pause, r to resume, q to quit.
    Status: Running

[HatCo.HatStore]
    orders-pending
        Country=Italy                                  6
        Country=Mexico                                 1
        Country=Spain                                  3
```

#### Best Practices

- Although the API allows any object to be used as the tag value, numeric types and strings are anticipated by collection tools. Other types may or may not be
  supported by a given collection tool.

- Beware of having very large or unbounded combinations of tag values being recorded in practice. Although the .NET API implementation can handle it, collection tools will
  likely allocate storage for metric data associated with each tag combination and this could become very large. For example it is fine if HatCo has 10 different
  hat colors and 25 hat sizes for up to 10*25=250 sales totals to track. However if HatCo added a 3rd tag that is a CustomerID for the sale and they sell to 100

>   million customers world-wide now there are now likely to be billions of different tag combinations being recorded. Most metric collection tools will either drop data

Reviewer 4: ```suggestion
  million customers worldwide, now there are now likely to be billions of different tag combinations being recorded. Most metric collection tools will either drop data
```

  to stay within technical limits or there can be large monetary costs to cover the data storage and processing. The implementation of each collection tool will determine

>   its limits but likely < 1K combinations for one instrument is safe, above 1K will need the collection tool to apply filtering or be engineered to operate at high scale.

Reviewer 4: ```suggestion
  its limits, but likely less than 1000 combinations for one instrument is safe. Anything above 1000 combinations will require the collection tool to apply filtering or be engineered to operate at high scale.
```

>   Histogram implementations tend to use far more memory than other metrics so safe limits could be 10-100x lower. If you anticipate large number of unique tag combinations

Reviewer 4: ```suggestion
  Histogram implementations tend to use far more memory than other metrics, so safe limits could be 10-100 times lower. If you anticipate large number of unique tag combinations,
```

>   then logs, transactional databases or big data processing systems may be more appropriate solutions to operate at the needed scale.

Reviewer 4: ```suggestion
  then logs, transactional databases, or big data processing systems may be more appropriate solutions to operate at the needed scale.
```

> - For instruments that will have very large numbers of tag combinations, prefer using a smaller storage type to help reduce memory overhead. For example storing the `short` for

Reviewer 4: ```suggestion
- For instruments that will have very large numbers of tag combinations, prefer using a smaller storage type to help reduce memory overhead. For example, storing the `short` for
```

>   a `Counter<short>` only occupies 2 bytes per tag combination whereas a `double` for `Counter<double>` occupies 8 bytes per tag combination.

Reviewer 4: ```suggestion
  a `Counter<short>` only occupies 2 bytes per tag combination, whereas a `double` for `Counter<double>` occupies 8 bytes per tag combination.
```

- Collection tools are encouraged to optimize for code that specifies the same set of tag names in the same order for each call to record measurements on the

>   same instrument. For high performance code that needs to call <xref:System.Diagnostics.Metrics.Counter`1.Add%2A> and <xref:System.Diagnostics.Metrics.Histogram`1.Record%2A>

Reviewer 4: ```suggestion
  same instrument. For high performance code that needs to call <xref:System.Diagnostics.Metrics.Counter%601.Add%2A> and <xref:System.Diagnostics.Metrics.Histogram%601.Record%2A>
```

>   frequently prefer using the same sequence of tag names for each call.

Reviewer 4: ```suggestion
  frequently, prefer using the same sequence of tag names for each call.
```

> - The .NET API is optimized to be allocation-free for <xref:System.Diagnostics.Metrics.Counter`1.Add%2A> and <xref:System.Diagnostics.Metrics.Histogram`1.Record%2A> calls

Reviewer 4: ```suggestion
- The .NET API is optimized to be allocation-free for <xref:System.Diagnostics.Metrics.Counter%601.Add%2A> and <xref:System.Diagnostics.Metrics.Histogram%601.Record%2A> calls
```

>   with three or fewer tags specified individually. To avoid allocations with larger numbers of tags, use the <xref:System.Diagnostics.TagList>. In general,

Reviewer 4: ```suggestion
  with three or fewer tags specified individually. To avoid allocations with larger numbers of tags, use <xref:System.Diagnostics.TagList>. In general,
```
